transport/tchannel/tchannel_integration_test.go

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tchannel_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/x/yarpctest"
+	"go.uber.org/yarpc/x/yarpctest/api"
+	"go.uber.org/yarpc/yarpcerrors"
+)
+
+func TestHandleResourceExhausted(t *testing.T) {
+	serviceName := "test-service"
+	procedureName := "test-procedure"
+	port := uint16(8000)
+
+	resourceExhaustedHandler := yarpctest.EchoHandler(api.UnaryInboundMiddlewareFunc(
+		func(context.Context, *transport.Request, transport.ResponseWriter, transport.UnaryHandler) error {
+			// eg: simulating a rate limiter that's reached its limit
```

Comment:
optional: the middleware implementation doesn't need the closure. Can move to an unexported top-level function.

```go
func resourceExhaustedMiddleware(context.Context, *transport.Request, transport.ResponseWriter, transport.UnaryHandler) error {
   ...
}
```

```go
handler := yarpctest.EchoHandler(api.UnaryInboundMiddlewareFunc(resourceExhaustedMiddleware))
```

Also, I'm not sure you need the echo handler at all. What's stopping you from implementing a `UnaryHandler` that returns this error without writing anything?

Reply:
Initially I thought I couldn't (without updating the testing library) since `api.UnaryHandlerFunc` does not match the `api.ProcOption` interface.

Did more digging into the testing library and found `types.UnaryHandler` does this. So 'll update the test and land it.